corehq/apps/hqwebapp/tests/test_forms.py

```diff
@@ -0,0 +1,82 @@
+from unittest.mock import patch, MagicMock, ANY
+from django.test import RequestFactory, SimpleTestCase
+from django.core.exceptions import ValidationError
+from corehq.apps.users.models import WebUser
+
+from ..forms import HQAuthenticationTokenForm, HQBackupTokenForm
+
+
+class HQAuthenticationTokenFormTests(SimpleTestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.setUp_mocks()
+
+    def setUp_mocks(self):
+        user_patcher = patch('corehq.apps.hqwebapp.forms.CouchUser.get_by_username',
+            new=MagicMock(side_effect=lambda x: self.user))
+        user_patcher.start()
+        self.addCleanup(user_patcher.stop)
+
+        clean_patcher = patch('corehq.apps.hqwebapp.forms.AuthenticationTokenForm.clean')
+        self.mocked_clean = clean_patcher.start()
+        self.mocked_clean.side_effect = ValidationError('Bad Token')
+        self.addCleanup(clean_patcher.stop)
+
+    def begin_login_attempt(self):
+        self.user = WebUser(username='test_user')
+        request = self.factory.post('/login')
+
+        return (self.user, request)
+
+    def create_form_with_invalid_token(self, user, request):
+        return HQAuthenticationTokenForm(user, 'device', request)
+
+    @patch('corehq.apps.hqwebapp.forms.user_login_failed')
+    def test_failed_authentication_sends_fully_formed_signal(self, mock_signal):
+        user, request = self.begin_login_attempt()
+        form = self.create_form_with_invalid_token(user, request)
+
+        with self.assertRaises(ValidationError):
+            form.clean()
+
+        expected_credentials = {'username': user.username}
+        mock_signal.send.assert_called_once_with(credentials=expected_credentials, request=request,
+            token_failure=True, sender=ANY)
+
+
+class HQBackupTokenFormTests(SimpleTestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.setUp_mocks()
+
+    def setUp_mocks(self):
+        user_patcher = patch('corehq.apps.hqwebapp.forms.CouchUser.get_by_username',
+            new=MagicMock(side_effect=lambda x: self.user))
+        user_patcher.start()
+        self.addCleanup(user_patcher.stop)
+
+        clean_patcher = patch('corehq.apps.hqwebapp.forms.BackupTokenForm.clean')
+        self.mocked_clean = clean_patcher.start()
+        self.mocked_clean.side_effect = ValidationError('Bad Token')
+        self.addCleanup(clean_patcher.stop)
+
+    def begin_login_attempt(self):
+        self.user = WebUser(username='test_user')
+        request = self.factory.post('/login')
+
+        return (self.user, request)
```

Comment:
There's a lot of duplication in this test class (comparing to the one before. What about combining the two classes and eliminating a lot of this duplicate code? Another option would be to refactor out the duplication into a common base class, but I don't like that option as much.

Reply:
This was intentional, as I don't try to stay DRY with test code. The token and backup token forms are functionally separate -- there is no reason their functionalities have to mirror each other. I guess you could argue, that at the point in time when their functionalities diverge, that would be the time to split the code. I thought of having a test hierarchy here, but I think we've gone over concerns on test hierarchies, and the indirection required turned me off of the idea.

Reply:
Fair enough. I'm not sure I agree 100% with "_I don't try to stay DRY with test code_" but I'm not going to stop you.